File: source/downloads/code/LeetCode/Python/Construct-Binary-Tree-from-Preorder-and-Inorder-Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """

        n = len(preorder)
        if n == 0:
            return None

        root = TreeNode(preorder[0])
        indexIn = inorder.index(preorder[0])
        leftL = indexIn
        root.left = self.buildTree(preorder[1:(1+leftL)], inorder[0:indexIn])
        root.right = self.buildTree(preorder[1+leftL:], inorder[indexIn+1,:])

        return root
        

File: downloads/code/LeetCode/Java/Construct-Binary-Tree-from-Preorder-and-Inorder-Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        int n = preorder.length;
        if (n==0)
            return null;
        return helper(preorder, 0, inorder, 0, n-1);
            
    }

    public TreeNode helper(int[] preorder,int ps,int[] inorder, int is, int ie){
        if(ps > preorder.length-1 || is > ie)
            return null;
        TreeNode node = new TreeNode(preorder[ps]);
        // find preorder[ps] in inorder: 左子树的长度
        int index=0;
        for(index=is; index<=ie;index++){
            if(inorder[index]==preorder[ps])
                break;
        }
        
        node.left = helper(preorder, ps+1,inorder,is,index-1);
        node.right = helper(preorder, ps+1+index-is,inorder,index+1, ie);
        return node; 
    
    } 
}
